File: PatronesEstructurales/facade.py
class CPU:
    def encender(self): print("CPU: encendida")
    def apagar(self):   print("CPU: apagada")

class RAM:
    def cargar_kernel(self): print("RAM: kernel cargado")
    def liberar(self):       print("RAM: memoria liberada")

class Disco:
    def leer_bootloader(self): print("Disco: bootloader leído")
    def guardar_estado(self):  print("Disco: estado guardado en disco")

class GPU:
    def activar(self, modo="normal"): print(f"GPU: activada (modo {modo})")
    def desactivar(self):             print("GPU: desactivada")

class SistemaOperativo:
    def iniciar(self):  print("SO: iniciado")
    def apagar(self):   print("SO: apagado")
    def modo_juego(self): print("SO: modo juego habilitado")
    def modo_ahorro(self): print("SO: modo ahorro habilitado")

# ------- Fachada -------
class Computador:
    def __init__(self):
        self.cpu = CPU()
        self.ram = RAM()
        self.disco = Disco()
        self.gpu = GPU()
        self.so  = SistemaOperativo()

    def encender(self):
        print(">> Encendiendo PC...")
        self.cpu.encender()
        self.disco.leer_bootloader()
        self.ram.cargar_kernel()
        self.so.iniciar()
        self.gpu.activar("normal")
        print(">> PC lista.\n")

    def apagar(self):
        print(">> Apagando PC...")
        self.so.apagar()
        self.disco.guardar_estado()
        self.gpu.desactivar()
        self.ram.liberar()
        self.cpu.apagar()
        print(">> PC apagada.\n")

    def modo_juego(self):
        print(">> Activando modo JUEGO 🎮")
        self.so.modo_juego()
        self.gpu.activar("alto rendimiento")
        print(">> FPS felices.\n")

    def modo_ahorro(self):
        print(">> Activando modo AHORRO 🔋")
        self.so.modo_ahorro()
        self.gpu.activar("eco")
        print(">> Batería agradecida.\n")

# ------- Uso -------
pc = Computador()
pc.encender()
pc.modo_juego()
pc.modo_ahorro()
pc.apagar()

File: PatronesEstructurales/Facade.java
package PatronesEstructurales;

class CPU {
    void encender() { System.out.println("CPU: encendida"); }
    void apagar() { System.out.println("CPU: apagada"); }
}

class RAM {
    void cargarKernel() { System.out.println("RAM: kernel cargado"); }
    void liberar() { System.out.println("RAM: memoria liberada"); }
}

class Disco {
    void leerBootloader() { System.out.println("Disco: bootloader leído"); }
    void guardarEstado() { System.out.println("Disco: estado guardado en disco"); }
}

class GPU {
    void activar(String modo) { System.out.println("GPU: activada (modo " + modo + ")"); }
    void desactivar() { System.out.println("GPU: desactivada"); }
}

class SistemaOperativo {
    void iniciar() { System.out.println("SO: iniciado"); }
    void apagar() { System.out.println("SO: apagado"); }
    void modoJuego() { System.out.println("SO: modo juego habilitado"); }
    void modoAhorro() { System.out.println("SO: modo ahorro habilitado"); }
}


class Computador {
    private final CPU cpu = new CPU();
    private final RAM ram = new RAM();
    private final Disco disco = new Disco();
    private final GPU gpu = new GPU();
    private final SistemaOperativo so = new SistemaOperativo();

    public void encender() {
        System.out.println("Encendiendo PC...");
        cpu.encender();
        disco.leerBootloader();
        ram.cargarKernel();
        so.iniciar();
        gpu.activar("normal");
        System.out.println("PC lista.\n");
    }

    public void apagar() {
        System.out.println("Apagando PC...");
        so.apagar();
        disco.guardarEstado();
        gpu.desactivar();
        ram.liberar();
        cpu.apagar();
        System.out.println("PC apagada.\n");
    }

    public void modoJuego() {
        System.out.println("Activando modo JUEGO 🎮");
        so.modoJuego();
        gpu.activar("alto rendimiento");
        System.out.println("FPS felices.\n");
    }

    public void modoAhorro() {
        System.out.println("Activando modo AHORRO 🔋");
        so.modoAhorro();
        gpu.activar("eco");
        System.out.println("Batería agradecida.\n");
    }
}


public class Facade {
    public static void main(String[] args) {
        Computador pc = new Computador();
        pc.encender();
        pc.modoJuego();
        pc.modoAhorro();
        pc.apagar();
    }
}
